## 38. 프로젝트 루트 확인

In [316]:
from pathlib import Path

project_root = Path.cwd()

if project_root.name == "notebooks":
    project_root = project_root.parent

data_dir = project_root / "data" / "raw"

print("프로젝트 루트:", project_root)
print("데이터 폴더:", data_dir)
print("데이터 폴더 존재:", data_dir.exists())

프로젝트 루트: c:\dev\ai-data-analysis
데이터 폴더: c:\dev\ai-data-analysis\data\raw
데이터 폴더 존재: True


## 39. 필수 파일 확인

In [317]:
required_files = [
    "customers.csv",
    "products.csv",
    "orders.csv",
    "order_items.csv",

]

for file_name in required_files:
    file_path = data_dir / file_name

    print(file_name, file_path.exists(), file_path.stat().st_size if file_path.exists() else 0)

customers.csv True 5641
products.csv True 4103
orders.csv True 11697
order_items.csv True 15375


In [318]:

import pandas as pd

customers = pd.read_csv(data_dir / "customers.csv")
products = pd.read_csv(data_dir / "products.csv")
orders = pd.read_csv(data_dir / "orders.csv")
order_items = pd.read_csv(data_dir / "order_items.csv")

## 41. DataFrame 사전 만들기

In [319]:
datasets = {
    "customers": customers,
    "products": products,
    "orders": orders,
    "order_items": order_items,
}

print(datasets)

{'customers':      customer_id name gender  age city signup_date
0              1  김수민      F   19   광주  2024-07-16
1              2  김정호      F   32   대구  2025-11-29
2              3  이경수      F   61   성남  2024-07-09
3              4  조영호      F   55   울산  2026-05-10
4              5  이예원      F   19   부산  2024-10-10
..           ...  ...    ...  ...  ...         ...
145          146  김숙자      M   61   성남  2026-01-19
146          147  이정남      M   19   부산  2025-03-10
147          148  오도현      M   29   고양  2026-07-12
148          149  김정자      M   20   부산  2024-11-15
149          150  조미영      M   40   대전  2025-12-31

[150 rows x 6 columns], 'products':     product_id product_name category   price
0            1  전자기기 상품 001     전자기기  160000
1            2    도서 상품 002       도서   34000
2            3  전자기기 상품 003     전자기기  152000
3            4  생활용품 상품 004     생활용품   70000
4            5    식품 상품 005       식품  186000
..         ...          ...      ...     ...
95          96  생활용품 상

## 42. 크기 요약

In [320]:
shape_summary = pd.DataFrame(
    [
        {
            "dataset": name,
            "rows": df.shape[0],
            "columns": df.shape[1],
        }
        for name, df in datasets.items()
    ]
)

print(shape_summary)

       dataset  rows  columns
0    customers   150        6
1     products   100        4
2       orders   301        5
3  order_items   766        5


## 43. 한 행의 의미 기록

### 데이터셋별 한 행의 의미
    항상 직접 작성하는거?
- customers: 고객 한 명
- products: 상품 한 개
- orders: 주문 한 건
- order_items: 주문에 포함된 상품 한 항목

## 44. 처음과 마지막 행 확인

In [321]:
for name, df in datasets.items():
    print(f"\n[{name}] 앞부분")
    display(df.head())
    print(f"[{name}] 뒷부분")
    display(df.tail())

    #왜 앞뒤 5개씩만 보여주는지?
    #빅데이터를 전체 다 확인 하는 것이 아닌 구조를 확인하기 위해서


[customers] 앞부분


,customer_id,name,gender,age,city,signup_date
0,1,김수민,F,19,광주,2024-07-16
1,2,김정호,F,32,대구,2025-11-29
2,3,이경수,F,61,성남,2024-07-09
3,4,조영호,F,55,울산,2026-05-10
4,5,이예원,F,19,부산,2024-10-10


[customers] 뒷부분


,customer_id,name,gender,age,city,signup_date
145,146,김숙자,M,61,성남,2026-01-19
146,147,이정남,M,19,부산,2025-03-10
147,148,오도현,M,29,고양,2026-07-12
148,149,김정자,M,20,부산,2024-11-15
149,150,조미영,M,40,대전,2025-12-31



[products] 앞부분


,product_id,product_name,category,price
0,1,전자기기 상품 001,전자기기,160000
1,2,도서 상품 002,도서,34000
2,3,전자기기 상품 003,전자기기,152000
3,4,생활용품 상품 004,생활용품,70000
4,5,식품 상품 005,식품,186000


[products] 뒷부분


,product_id,product_name,category,price
95,96,생활용품 상품 096,생활용품,112000
96,97,전자기기 상품 097,전자기기,154000
97,98,스포츠 상품 098,스포츠,10000
98,99,뷰티 상품 099,뷰티,200000
99,100,도서 상품 100,도서,102000



[orders] 앞부분


,order_id,customer_id,order_date,payment_method,order_status
0,1,123,2026-06-03,card,completed
1,2,77,2025-08-19,naver_pay,cancelled
2,3,138,2025-12-16,bank_transfer,cancelled
3,4,57,2026-02-26,kakao_pay,cancelled
4,5,125,2026-01-17,card,cancelled


[orders] 뒷부분


,order_id,customer_id,order_date,payment_method,order_status
296,297,22,2026-02-11,bank_transfer,completed
297,298,64,2026-01-21,naver_pay,cancelled
298,299,135,2025-09-04,bank_transfer,completed
299,300,96,2026-03-24,kakao_pay,cancelled
300,321,300,2026-01넌24,kakao_pay,cancelled



[order_items] 앞부분


,order_item_id,order_id,product_id,quantity,unit_price
0,1,1,100,3,102000
1,2,1,87,5,25000
2,3,1,7,3,142000
3,4,1,9,3,193000
4,5,2,72,4,189000


[order_items] 뒷부분


,order_item_id,order_id,product_id,quantity,unit_price
761,762,299,8,2,189000
762,763,299,12,4,175000
763,764,300,59,5,32000
764,789,320,67,5,32000
765,763,299,108,4,175000


## 45. 컬럼명 확인

In [322]:
for name, df in datasets.items():
    print(name, df.columns.tolist())

customers ['customer_id', 'name', 'gender', 'age', 'city', 'signup_date']
products ['product_id', 'product_name', 'category', 'price']
orders ['order_id', 'customer_id', 'order_date', 'payment_method', 'order_status']
order_items ['order_item_id', 'order_id', 'product_id', 'quantity', 'unit_price']


## 46. 컬럼명 공백 점검

In [323]:

for name, df in datasets.items():

    suspicious = [
        col for col in df.columns
        if col != col.strip()
    ]

    print(name, suspicious)

customers []
products []
orders []
order_items []


## 47. info() 실행 
    info()와 dtypes()는 매우 중요하다?

In [324]:
for name, df in datasets.items():
    print(f"\n===== {name} =====")
    df.info()


===== customers =====
<class 'pandas.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype
---  ------       --------------  -----
 0   customer_id  150 non-null    int64
 1   name         150 non-null    str  
 2   gender       150 non-null    str  
 3   age          150 non-null    int64
 4   city         150 non-null    str  
 5   signup_date  150 non-null    str  
dtypes: int64(2), str(4)
memory usage: 7.2 KB

===== products =====
<class 'pandas.DataFrame'>
RangeIndex: 100 entries, 0 to 99
Data columns (total 4 columns):
 #   Column        Non-Null Count  Dtype
---  ------        --------------  -----
 0   product_id    100 non-null    int64
 1   product_name  100 non-null    str  
 2   category      100 non-null    str  
 3   price         100 non-null    int64
dtypes: int64(2), str(2)
memory usage: 3.3 KB

===== orders =====
<class 'pandas.DataFrame'>
RangeIndex: 301 entries, 0 to 300
Data columns (total 5 columns):

## 48. dtype 요약표

In [325]:

dtype_rows = []

for name, df in datasets.items():
    for column, dtype in df.dtypes.items():
        dtype_rows.append(
            {
                "dataset": name,
                "column": column,
                "dtype": str(dtype),
            }
        )
dtype_summary = pd.DataFrame(dtype_rows)

print(dtype_summary)

        dataset          column  dtype
0     customers     customer_id  int64
1     customers            name    str
2     customers          gender    str
3     customers             age  int64
4     customers            city    str
5     customers     signup_date    str
6      products      product_id  int64
7      products    product_name    str
8      products        category    str
9      products           price  int64
10       orders        order_id  int64
11       orders     customer_id  int64
12       orders      order_date    str
13       orders  payment_method    str
14       orders    order_status    str
15  order_items   order_item_id  int64
16  order_items        order_id  int64
17  order_items      product_id  int64
18  order_items        quantity  int64
19  order_items      unit_price  int64


## 49. 날짜형 후보 확인

In [326]:
date_columns = {
    "customers": ["signup_date"],
    "orders": ["order_date"],
}

for dataset_name, columns in date_columns.items():
    df = datasets[dataset_name]
    for column in columns:
        converted = pd.to_datetime(
            df[column],
            errors="coerce"
        )

        print(
            dataset_name,
            column,
            "원본 dtype:", df[column].dtype,
            "변환 실패:", converted.isna().sum()
        )

customers signup_date 원본 dtype: str 변환 실패: 0
orders order_date 원본 dtype: str 변환 실패: 1


## 50. 결측치 개수

In [327]:
for name, df in datasets.items():
    print(f"\n[{name}]")
    display(df.isna().sum().to_frame("missing_count"))


[customers]


,missing_count
customer_id,0
name,0
gender,0
age,0
city,0
signup_date,0



[products]


,missing_count
product_id,0
product_name,0
category,0
price,0



[orders]


,missing_count
order_id,0
customer_id,0
order_date,0
payment_method,0
order_status,0



[order_items]


,missing_count
order_item_id,0
order_id,0
product_id,0
quantity,0
unit_price,0


## 51. 결측치 비율

In [328]:
missing_summaries = []

for name, df in datasets.items():
    summary = pd.DataFrame(
        {
            "missing_count": df.isna().sum(),
            "missing_rate_pct": (
                df.isna().mean() * 100
            ).round(2),
        }
    )
    summary.insert(0, "dataset", name)
    summary.insert(1, "column", summary.index)
    missing_summaries.append(summary.reset_index(drop=True))

missing_summary = pd.concat(
    missing_summaries,
    ignore_index=True
)

print(missing_summary)

        dataset          column  missing_count  missing_rate_pct
0     customers     customer_id              0               0.0
1     customers            name              0               0.0
2     customers          gender              0               0.0
3     customers             age              0               0.0
4     customers            city              0               0.0
5     customers     signup_date              0               0.0
6      products      product_id              0               0.0
7      products    product_name              0               0.0
8      products        category              0               0.0
9      products           price              0               0.0
10       orders        order_id              0               0.0
11       orders     customer_id              0               0.0
12       orders      order_date              0               0.0
13       orders  payment_method              0               0.0
14       orders    order_

## 52. 전체 행 중복

In [329]:
for name, df in datasets.items():
    print(name, "전체 중복:", df.duplicated().sum())

#duplicated()는 데이터프레임에서 중복된 행을 찾는 메서드입니다. 
# 이 메서드는 각 행이 이전에 나타난 적이 있는지 확인하고, 
# 중복된 행이면 True를 반환하고, 그렇지 않으면 False를 반환합니다.

customers 전체 중복: 0
products 전체 중복: 0
orders 전체 중복: 0
order_items 전체 중복: 0


## 53. 주요 ID 점검

In [330]:
primary_keys = {
    "customers": "customer_id",
    "products": "product_id",
    "orders": "order_id",
    "order_items": "order_item_id",
}
id_check_rows = []

for name, key in primary_keys.items():
    df = datasets[name]

    id_check_rows.append(
        {
            "dataset": name,
            "key": key,
            "missing": df[key].isna().sum(),
            "duplicates": df[key].duplicated().sum(),
            "unique": df[key].nunique(),
            "rows": len(df),
        }
    )
id_check_summary = pd.DataFrame(id_check_rows)

print(id_check_summary)

       dataset            key  missing  duplicates  unique  rows
0    customers    customer_id        0           0     150   150
1     products     product_id        0           0     100   100
2       orders       order_id        0           0     301   301
3  order_items  order_item_id        0           1     765   766


## 데이터 품질 발견 기록

| 데이터셋 | 컬럼 | 문제 종류 | 건수 | 쉬운 설명 | 처리 여부 |
|---|---|---|---|---|---|
| 고객(customers) | 전체 | 빈 칸 / 중복된 행 | 0 | 비어있는 값이나 똑같이 겹치는 행이 없음 | N/A |
| 고객(customers) | 나이 | 말이 안 되는 값 (예: 나이가 음수거나 200살) | 0 | 나이는 19~69세로 정상 범위 | N/A |
| 고객(customers) | 성별 | 값이 통일되지 않음 (예: F, 여자, female 섞임) | 0 | F/M 두 가지로만 통일되어 있음 | N/A |
| 고객(customers) | 가입일 | 날짜 형식 오류 / 미래 날짜 | 0 | 날짜 형식 문제없고, 미래 날짜(아직 안 온 날)도 없음 | N/A |
| 주문(orders) | 전체 | 빈 칸 / 중복된 행 | 0 | 비어있는 값이나 겹치는 행이 없음 | N/A |
| 주문(orders) | 고객 번호 | 존재하지 않는 고객 연결 | 0 | 주문에 적힌 고객이 실제 고객 목록에 다 있음 | N/A |
| 주문(orders) | 주문일 | 날짜 형식 오류 / 미래 날짜 | 0 | 날짜 형식 문제없고, 미래 날짜도 없음 | N/A |
| 상품(products) | 가격 | 0원 이하 이상한 가격 | 0 | 가격은 5,000원~200,000원 사이로 정상 | N/A |
| 주문상품(order_items) | 주문번호 / 상품번호 | 존재하지 않는 주문·상품과 연결 | 0 | 실제 주문·상품 목록에 다 존재함 | N/A |
| 주문상품(order_items) | 수량, 단가 | 0 이하 이상한 값 | 0 | 수량·단가 모두 정상적인 양수 값 | N/A |
| 주문상품(order_items) | 단가 | 상품 원래 가격과 다름 | 0 | 주문서 단가와 상품 목록 가격이 정확히 일치 | N/A |

## 55. 숫자형 기본 통계

In [331]:
# 55. 숫자형 기본 통계

print(customers[["age"]].describe())
print(products[["price"]].describe())
print(order_items[["quantity", "unit_price"]].describe())

#quantity는 수량, unit_price는 단가를 의미합니다.
#describe() 메서드는 count, mean, std, min, 25%, 50%, 75%, max 등의 통계 정보를 제공합니다.

              age
count  150.000000
mean    42.086667
std     15.613166
min     19.000000
25%     29.000000
50%     40.000000
75%     57.000000
max     69.000000
               price
count     100.000000
mean   110040.000000
std     56433.910574
min      5000.000000
25%     65750.000000
50%    112000.000000
75%    161000.000000
max    200000.000000
         quantity     unit_price
count  766.000000     766.000000
mean     3.057441  108548.302872
std      1.411194   57040.081227
min      1.000000    5000.000000
25%      2.000000   62000.000000
50%      3.000000  111000.000000
75%      4.000000  161750.000000
max      5.000000  200000.000000


In [332]:
#56. 범주형 값 확인

category_columns = {
    "customers": ["gender", "city"],
    "products": ["category"],
    "orders": ["payment_method", "order_status"],
}

for dataset_name, columns in category_columns.items():
    df = datasets[dataset_name]
    for column in columns:
        print(f"\n[{dataset_name}.{column}]")

        display(
            df[column]
            .value_counts(dropna=False)
            .to_frame("count")
        )


[customers.gender]


,count
gender,
F,84
M,66



[customers.city]


,count
city,
성남,21
광주,17
부산,16
대구,15
서울,15
울산,14
인천,14
대전,14
수원,13



[products.category]


,count
category,
스포츠,19
전자기기,17
생활용품,16
뷰티,16
도서,14
패션,11
식품,7



[orders.payment_method]


,count
payment_method,
kakao_pay,80
naver_pay,77
bank_transfer,74
card,70



[orders.order_status]


,count
order_status,
completed,184
cancelled,65
refunded,52


In [333]:
# 57. 날짜 범위 확인

signup_dates = pd.to_datetime(
    customers["signup_date"],
    errors="coerce"
)
order_dates = pd.to_datetime(
    orders["order_date"],
    errors="coerce"
)
print("가입일 변환 실패:", signup_dates.isna().sum())
print("가입일 범위:", signup_dates.min(), "~", signup_dates.max())
print("주문일 변환 실패:", order_dates.isna().sum())
print("주문일 범위:", order_dates.min(), "~", order_dates.max())

#isna()는 결측값을 확인하는 메서드입니다.

가입일 변환 실패: 0
가입일 범위: 2023-08-10 00:00:00 ~ 2026-07-24 00:00:00
주문일 변환 실패: 1
주문일 범위: 2025-08-05 00:00:00 ~ 2026-08-04 00:00:00


## 58. 날짜 관계 점검 준비

주문일과 고객 가입일을 직접 비교하려면 두 데이터를 연결해야 합니다.

이번 장에서는 구조와 키가 연결 가능한지 확인하고, 

실제 병합과 분석은 이후 장에서 본격적으로 진행합니다.

In [334]:
# 59. 참조하지 못한 고객 ID

unknown_customer_ids = sorted(
    set(orders["customer_id"].dropna())
    - set(customers["customer_id"].dropna())
)
print("존재하지 않는 customer_id:", unknown_customer_ids)



존재하지 않는 customer_id: [300]


In [335]:
# 60. 참조하지 못한 주문 ID

unknown_order_ids = sorted(
    set(order_items["order_id"].dropna())
    - set(orders["order_id"].dropna())
)
print("존재하지 않는 order_id:", unknown_order_ids)

존재하지 않는 order_id: [320]


In [336]:
# 61. 참조하지 못한 상품 ID

unknown_product_ids = sorted(
    set(order_items["product_id"].dropna())
    - set(products["product_id"].dropna())
)
print("존재하지 않는 product_id:", unknown_product_ids)

존재하지 않는 product_id: [108]


In [337]:
# 62. 반복 점검 함수

def inspect_dataframe(
    name: str,
    df: pd.DataFrame,
    primary_key: str | None = None,
) -> dict:
    result = {
        "dataset": name,
        "rows": df.shape[0],
        "columns": df.shape[1],
        "column_names": df.columns.tolist(),
        "dtypes": {
            column: str(dtype)
            for column, dtype in df.dtypes.items()
        },
        "missing_total": int(df.isna().sum().sum()),
        "duplicate_rows": int(df.duplicated().sum()),
    }
    if primary_key is not None:
        result["primary_key"] = primary_key
        result["key_missing"] = int(
            df[primary_key].isna().sum()
        )
        result["key_duplicates"] = int(
            df[primary_key].duplicated().sum()
        )
        result["key_unique"] = int(
            df[primary_key].nunique()
        )
    return result

In [338]:
inspection_results = []

for name, df in datasets.items():
    inspection_results.append(
        inspect_dataframe(
            name,
            df,
            primary_keys[name],
        )
    )
print(inspection_results)

[{'dataset': 'customers', 'rows': 150, 'columns': 6, 'column_names': ['customer_id', 'name', 'gender', 'age', 'city', 'signup_date'], 'dtypes': {'customer_id': 'int64', 'name': 'str', 'gender': 'str', 'age': 'int64', 'city': 'str', 'signup_date': 'str'}, 'missing_total': 0, 'duplicate_rows': 0, 'primary_key': 'customer_id', 'key_missing': 0, 'key_duplicates': 0, 'key_unique': 150}, {'dataset': 'products', 'rows': 100, 'columns': 4, 'column_names': ['product_id', 'product_name', 'category', 'price'], 'dtypes': {'product_id': 'int64', 'product_name': 'str', 'category': 'str', 'price': 'int64'}, 'missing_total': 0, 'duplicate_rows': 0, 'primary_key': 'product_id', 'key_missing': 0, 'key_duplicates': 0, 'key_unique': 100}, {'dataset': 'orders', 'rows': 301, 'columns': 5, 'column_names': ['order_id', 'customer_id', 'order_date', 'payment_method', 'order_status'], 'dtypes': {'order_id': 'int64', 'customer_id': 'int64', 'order_date': 'str', 'payment_method': 'str', 'order_status': 'str'}, 'mi

In [339]:
# 61. 참조하지 못한 상품 ID

unknown_product_ids = sorted(
    set(order_items["product_id"].dropna())
    - set(products["product_id"].dropna())
)
print("존재하지 않는 product_id:", unknown_product_ids)

존재하지 않는 product_id: [108]


In [340]:
inspection_results = []
for name, df in datasets.items():
    inspection_results.append(
        inspect_dataframe(
            name,
            df,
            primary_keys[name],
        )
    )

inspection_results

[{'dataset': 'customers',
  'rows': 150,
  'columns': 6,
  'column_names': ['customer_id',
   'name',
   'gender',
   'age',
   'city',
   'signup_date'],
  'dtypes': {'customer_id': 'int64',
   'name': 'str',
   'gender': 'str',
   'age': 'int64',
   'city': 'str',
   'signup_date': 'str'},
  'missing_total': 0,
  'duplicate_rows': 0,
  'primary_key': 'customer_id',
  'key_missing': 0,
  'key_duplicates': 0,
  'key_unique': 150},
 {'dataset': 'products',
  'rows': 100,
  'columns': 4,
  'column_names': ['product_id', 'product_name', 'category', 'price'],
  'dtypes': {'product_id': 'int64',
   'product_name': 'str',
   'category': 'str',
   'price': 'int64'},
  'missing_total': 0,
  'duplicate_rows': 0,
  'primary_key': 'product_id',
  'key_missing': 0,
  'key_duplicates': 0,
  'key_unique': 100},
 {'dataset': 'orders',
  'rows': 301,
  'columns': 5,
  'column_names': ['order_id',
   'customer_id',
   'order_date',
   'payment_method',
   'order_status'],
  'dtypes': {'order_id': 'int6

In [341]:
# 62. 반복 점검 함수

def inspect_dataframe(
    name: str,
    df: pd.DataFrame,
    primary_key: str | None = None,
) -> dict:
    result = {
        "dataset": name,
        "rows": df.shape[0],
        "columns": df.shape[1],
        "column_names": df.columns.tolist(),
        "dtypes": {
            column: str(dtype)
            for column, dtype in df.dtypes.items()
        },
        "missing_total": int(df.isna().sum().sum()),
        "duplicate_rows": int(df.duplicated().sum()),
    }
    if primary_key is not None:
        result["primary_key"] = primary_key
        result["key_missing"] = int(
            df[primary_key].isna().sum()
        )
        result["key_duplicates"] = int(
            df[primary_key].duplicated().sum()
        )
        result["key_unique"] = int(
            df[primary_key].nunique()
        )
    return result

In [343]:
# 63. LLM에 전달할 안전한 구조 요약

safe_schema_summary = {
    "datasets": inspection_results,
    "relationships": {
        "orders.customer_id": {
            "references": "customers.customer_id",
            "unknown_count": len(unknown_customer_ids),
        },
        "order_items.order_id": {
            "references": "orders.order_id",
            "unknown_count": len(unknown_order_ids),
        },
        "order_items.product_id": {
            "references": "products.product_id",
            "unknown_count": len(unknown_product_ids),
        },
    },
}

print(safe_schema_summary)

{'datasets': [{'dataset': 'customers', 'rows': 150, 'columns': 6, 'column_names': ['customer_id', 'name', 'gender', 'age', 'city', 'signup_date'], 'dtypes': {'customer_id': 'int64', 'name': 'str', 'gender': 'str', 'age': 'int64', 'city': 'str', 'signup_date': 'str'}, 'missing_total': 0, 'duplicate_rows': 0, 'primary_key': 'customer_id', 'key_missing': 0, 'key_duplicates': 0, 'key_unique': 150}, {'dataset': 'products', 'rows': 100, 'columns': 4, 'column_names': ['product_id', 'product_name', 'category', 'price'], 'dtypes': {'product_id': 'int64', 'product_name': 'str', 'category': 'str', 'price': 'int64'}, 'missing_total': 0, 'duplicate_rows': 0, 'primary_key': 'product_id', 'key_missing': 0, 'key_duplicates': 0, 'key_unique': 100}, {'dataset': 'orders', 'rows': 301, 'columns': 5, 'column_names': ['order_id', 'customer_id', 'order_date', 'payment_method', 'order_status'], 'dtypes': {'order_id': 'int64', 'customer_id': 'int64', 'order_date': 'str', 'payment_method': 'str', 'order_status'

## D. 실습 결과 기록 양식

## 65. LLM답변 검증표

| 검증 항목 | 확인 내용 | 검증 결과 |
|---|---|---|
| 컬럼 존재 | 실제 컬럼명과 일치하는가? | 일치 |
| 한 행의 의미 | 데이터셋 역할과 일치하는가? | 일치 |
| ID 해석 | 식별자를 수치 변수로 해석하지 않았는가? | 미해당 (식별자로만 사용) |
| 중복 해석 | 반복 가능한 외래 키를 오류로 판단하지 않았는가? | 정상 구분함 |
| 결측 해석 | 업무 의미 확인 없이 삭제를 권하지 않았는가? | 삭제 미권고 |
| 관계 | 실제 연결 키와 일치하는가? | 일치 |
| 개인정보 | 추가 원본 행 제공을 요구하지 않았는가? | 요구 없음 |

## 데이터 첫인상 보고서

## 1. 데이터 파일

| 데이터셋 | 한 행의 의미 | 행 | 열 | 주요 ID |
|---|---|---:|---:|---|
| customers | 고객 1명의 회원 정보 | 150 | 6 | customer_id |
| products | 상품 카탈로그의 상품 1개 | 100 | 4 | product_id |
| orders | 주문 1건 (주문 헤더) | 301 | 5 | order_id |
| order_items | 한 주문에 담긴 상품 1줄 | 766 | 5 | order_item_id |

## 2. 컬럼과 타입

| 데이터셋 | 컬럼 | 타입 |
|---|---|---|
| customers | customer_id / name / gender / age / city / signup_date | int64 / str / str / int64 / str / str |
| products | product_id / product_name / category / price | int64 / str / str / int64 |
| orders | order_id / customer_id / order_date / payment_method / order_status | int64 / int64 / str / str / str |
| order_items | order_item_id / order_id / product_id / quantity / unit_price | int64 / int64 / int64 / int64 / int64 |

## 3. 결측치

| 데이터셋 | 결측치 (missing_total) |
|---|---|
| customers | 0건 |
| products | 0건 |
| orders | 0건 |
| order_items | 0건 |

## 4. 전체 중복

| 데이터셋 | 전체 행 기준 중복 (duplicate_rows) |
|---|---|
| customers | 0건 |
| products | 0건 |
| orders | 0건 |
| order_items | 0건 |

## 5. 주요 ID 결측·중복

| 데이터셋 | 주요 ID | 결측(key_missing) | 중복(key_duplicates) | 고유값 수(key_unique) |
|---|---|---:|---:|---:|
| customers | customer_id | 0 | 0 | 150 |
| products | product_id | 0 | 0 | 100 |
| orders | order_id | 0 | 0 | 301 |
| order_items | order_item_id | 0 | **1** | 765 |

## 6. 숫자형 범위

제공된 요약 정보에 숫자형 범위(최소·최대·이상값) 항목이 포함되어 있지 않음. 원본 파일에서 별도 확인 필요.

## 7. 범주형 값

제공된 요약 정보에 범주형 값 목록·분포 항목이 포함되어 있지 않음. 원본 파일에서 별도 확인 필요.

## 8. 날짜 범위와 변환 실패

제공된 요약 정보에 날짜 범위·형식 변환 실패 항목이 포함되어 있지 않음. 원본 파일에서 별도 확인 필요.

## 9. 파일 간 참조 무결성

| 관계 | 존재하지 않는 값의 개수 (unknown_count) |
|---|---:|
| orders.customer_id → customers.customer_id | 1 |
| order_items.order_id → orders.order_id | 1 |
| order_items.product_id → products.product_id | 1 |

## 10. 발견한 문제

1. `order_items`의 주요 ID(`order_item_id`)에 중복 1건 존재 (key_duplicates=1)
2. `orders.customer_id`가 참조하는 `customers.customer_id` 중 존재하지 않는 값 1건
3. `order_items.order_id`가 참조하는 `orders.order_id` 중 존재하지 않는 값 1건
4. `order_items.product_id`가 참조하는 `products.product_id` 중 존재하지 않는 값 1건

## 11. 즉시 처리하지 않은 이유

- 이 요약 정보만으로는 **어떤 행(row)**이 문제인지, 구체적으로 어떤 값이 중복·불일치하는지 알 수 없음
- 원인 파악(입력 오류인지, 참조 대상 테이블의 누락인지) 없이 건수만 보고 삭제·수정하면 실제 데이터를 잘못 훼손할 위험이 있음
- 결측치가 아니라 **중복/참조 오류**이므로 일반적인 결측치 처리 방식(삭제 등)을 그대로 적용할 수 없음

## 12. 다음 전처리 단계에서 결정할 내용

- 원본 파일을 직접 열어 `order_items` 중복 ID(1건)의 실제 행 2개를 비교해 어느 것이 올바른 데이터인지 결정
- 참조무결성 위반 3건(주문↔고객, 주문상세↔주문, 주문상세↔상품)의 실제 값을 확인해 원인(오탈자 vs 마스터 테이블 누락)을 파악
- 숫자형 범위·범주형 값·날짜 형식은 이번 요약에 없으므로, 별도로 원본 파일을 검사할지 여부 결정
- 위 문제 행들을 분석에 포함할지, 제외할지, 별도 플래그로 표시할지 결정